### ISSUE TITLE
switch-exhaustiveness-check: typeof

### ISSUE THREAD

**Issue by ilyub** [typescript-eslint/typescript-eslint#3453]
**Repro**

```ts
function f(x: number | string): void {
    switch (typeof x) {
      case "number":
        ...
      case "string":
        ...
    }
  }
```

**Expected Result**

Expecting this code to pass

**Actual Result**

Getting "Switch is not exhaustive" warning.

**Additional Info**

I know that typeof has "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function", which evidently triggers error. However, in the above example it is obvious that switch is exhaustive.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.25.0` |
| `@typescript-eslint/parser`        | `4.25.0` |
| `TypeScript`                       | `4.3.2` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `6.14.13` |

**Comment by bradzacher** [comment]
TypeScript always tells us the type of `typeof x` includes every possible string:
![image](https://user-images.githubusercontent.com/7462525/119902780-907bca80-befc-11eb-836f-2379271d574d.png)

But then it has special logic built into it which leverages the actual, underlying type of `x` to properly narrow the type of `x`:
![image](https://user-images.githubusercontent.com/7462525/119902914-be610f00-befc-11eb-8faa-551e8bcb4de7.png)

I'm not sure if this is something we can handle - it will require a decent chunk of work I suspect.

**Comment by ilyub** [comment]
I have checked the code for this rule and I see this snippet:
```ts
function checkSwitchExhaustive(node: TSESTree.SwitchStatement): void {
      const discriminantType = getNodeType(node.discriminant);
```

I am sure this code can be adjusted something like this:
```ts
function checkSwitchExhaustive(node: TSESTree.SwitchStatement): void {
  const discriminantType = node.discriminant === _typeof_operator_
    ? getNodeType(node.discriminant._argument_of_typeof_operator)
    : getNodeType(node.discriminant);
```
In other words, if we see "switch (typeof x)" then we take type of "x", instead of taking type of "typeof x". This looks like an easy fix and it will solve the problem for this pattern:
```ts
switch (typeof x) {
  ...
}
```

We can ignore patterns like this:
```ts
const type = typeof x;
switch (x) {
  ...
}
```
In fact, there is a reason why the first pattern is important. Only the first pattern serves as type guard:
```ts
switch (typeof x) {
  case "string":
    x.toLowerCase(); // Typescript compiler narrows the type of "x" to "string"
}
```
The second pattern:
```ts
const type = typeof x;
switch (x) {
  case "string":
    x.toLowerCase(); // ERROR!!! Typescript does not narrow type.
}
```
does not narrow type. So, it is of much less use and can be ignored.

The problem is that currently I have to add redundant default:
```ts
switch (typeof x) {
  case "string":
    x.toLowerCase(); // Typescript compiler narrows the type of "x" to "string"
  default:
    // Do nothing
}
```

Redundant default has two problems:

1. It can't be reached by tests. So, it lowers test coverage.
2. If the type of "x" variable changes e.g. from "string | number" to "string |  number | object", programmer may forget to write the code to handle new case. Without redundant "default" linter would complain that switch is not exaustive and this would let you know that you need to add case for "object". With redundant "default" linter will give no hint.

So, I suggest to make a limited fix only for this most important pattern: "switch (typeof x)". I hope this is possible.

**Comment by bradzacher** [comment]
I like the enthusiasm! But it's not quite as simple as you make it out to be.

The rule currently works like this:
1) get all the types in the test's union.
2) get the type of each case.
3) assert that all types in the union exist as a case argument, or that a `default` exists.

However, `typeof x` returns a string union, but `x`'s type is not those types.
This means that step 3 above will fail, because `number` and `'number'` are two different types.

So instead the rule would have to be taught to understand how any given type maps to a specific member of the `typeof` union, so that it can be checked appropriately.
This isn't super simple to do - it'll take a decent chunk of work to handle and test all cases.
